internal/bgp/bgpfrr/frr.go

```diff
@@ -0,0 +1,178 @@
+package bgpfrr
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"go.universe.tf/metallb/internal/bgp"
+)
+
+// As the MetalLB controller should handle messages synchronously, there should
+// no need to lock this data structure. TODO: confirm this.
+type frrState struct {
+	sessions map[string]*Session
+}
+
+var state *frrState
+
+func init() {
+	state = &frrState{
+		sessions: map[string]*Session{},
+	}
+}
+
+type Session struct {
+	myASN      uint32
+	routerID   net.IP // May be nil, meaning "derive from context"
+	myNode     string
+	addr       string
+	srcAddr    net.IP
+	asn        uint32
+	holdTime   time.Duration
+	logger     log.Logger
+	password   string
+	advertised map[string]*bgp.Advertisement
+}
+
+// sessionName() defines the format of the key of the 'sessions' map in
+// the 'frrState' struct.
+func sessionName(routerID string, myAsn uint32, addr string, asn uint32) string {
+	return fmt.Sprintf("%d@%s-%d@%s", asn, addr, myAsn, routerID)
+}
+
+func validate(adv *bgp.Advertisement) error {
+	if adv.Prefix.IP.To4() != nil {
+		if adv.NextHop != nil && adv.NextHop.To4() == nil {
+			return fmt.Errorf("next-hop must be IPv4, got %q", adv.NextHop)
+		}
+	} else if adv.Prefix.IP.To16() != nil {
+		if adv.NextHop != nil && adv.NextHop.To16() == nil {
+			return fmt.Errorf("next-hop must be IPv6, got %q", adv.NextHop)
+		}
+	} else {
+		return fmt.Errorf("unable to validate IP address")
+	}
+
+	if len(adv.Communities) > 63 {
+		return fmt.Errorf("max supported communities is 63, got %d", len(adv.Communities))
+	}
+	return nil
+}
+
+func (s *Session) Set(advs ...*bgp.Advertisement) error {
```

Comment:
Another thing to note is, we are probably calling this (and New) multiple times whenever we receive a new configuration / a loadbalancer event (https://github.com/metallb/metallb/blob/717fd1dfae2852ce95f6e05c580d177d0d44c35a/speaker/bgp_controller.go#L258)
Would it make sense to extend the interface with a `commit` sort of mechanism?
I know this is falling in the land of early optimization, but the file will blink (at least) one time per peer.

Reply:
Yeah, I think it would but as you mention - early optimization. IIRC, the reloader will hide some of this as it only applies the differences but we would need to confirm and test.

I would suggest we address this in a follow-up PR. Let's get something working and passing a BGP CI first and we can make those updates later? What do you think?

Reply:
Sounds good to me

Reply:
On second thoughts, it didn't seem so hard to do. I did an implementation of this. I added it as two separate commits in this PR. Maybe have a look at it and if you think it is worthwhile.

Reply:
Looks great, and I think it will save us from some headaches!